### PR TITLE
[xchain-eth] Refactor fee handling

### PR DESCRIPTION
**Pull request by veado**
- [x] `xchain-client`: Change `getFee` to provide optional parameter
- [x] `xchain-eth`: Update `getFee` in `xchain-eth`
- [x] `xchain-eth`: Add `estimateGasLimit`, `estimateGasLimits`, `estimateFeesWithGasPricesAndLimits`
- [x] `xchain-eth`: Update tests 
- [x] Prepare `xchainjs/xchain-client@0.4.0`
- [x] Prepare `xchainjs/xchain-eth@0.9.0`

### CODE REVIEW

File: packages/xchain-ethereum/src/utils.ts [hunk 1]
```diff
@@ -1,12 +1,33 @@
...
+  FeesWithGasPricesAndLimits,
+} from './types'
+import { ethers, BigNumber } from 'ethers'
+import { parseUnits } from 'ethers/lib/utils'
 
 export const ETH_DECIMAL = 18
-export const DEFAULT_GASLIMIT = 63000
 export const ETHPLORER_FREEKEY = 'freekey'
 
+// from https://github.com/MetaMask/metamask-extension/blob/ee205b893fe61dc4736efc576e0663189a9d23da/ui/app/pages/send/send.constants.js#L39
```
Comment (veado) [inline packages/xchain-ethereum/src/utils.ts:24]: [nitpick] ... and based on recommendations of https://ethgasstation.info/blog/gas-limit/

File: packages/xchain-ethereum/src/utils.ts [hunk 2]
```diff
@@ -148,19 +169,64 @@ export const getTxFromEthTransaction = (tx: ETHTransactionInfo): Tx => {
...
  */
-export const getDefaultFees = (): Fees => {
+export const getFee = ({ gasPrice, gasLimit }: { gasPrice: BaseAmount; gasLimit: BigNumber }) =>
+  baseAmount(gasPrice.amount().multipliedBy(gasLimit.toString()))
+
+export const estimateDefaultFeesWithGasPricesAndLimits = (asset?: Asset): FeesWithGasPricesAndLimits => {
+  const gasPrices = {
+    average: baseAmount(parseUnits('30', 'gwei').toString(), ETH_DECIMAL),
+    fast: baseAmount(parseUnits('35', 'gwei').toString(), ETH_DECIMAL),
+    fastest: baseAmount(parseUnits('39', 'gwei').toString(), ETH_DECIMAL),
```
Comment (veado) [inline packages/xchain-ethereum/src/utils.ts:183]: Where are values for default (`30`, `35`, `39`) based from?

File: packages/xchain-ethereum/src/client.ts [hunk 3]
```diff
@@ -566,23 +580,70 @@ export default class Client implements XChainClient, EthereumClient {
...
+    ]).then(([fast, fastest, average]) => ({
+      average,
+      fast,
+      fastest,
+    }))
+  }
+
+  estimateFeesWithGasPricesAndLimits = async (params: FeesParams): Promise<FeesWithGasPricesAndLimits> => {
+    // gas prices
+    const gasPrices = await this.estimateGasPrices()
```
Comment (veado) [inline packages/xchain-ethereum/src/client.ts:618]: Handle failing of `estimateGasPrices`

File: packages/xchain-ethereum/src/client.ts [hunk 4]
```diff
@@ -566,23 +580,70 @@ export default class Client implements XChainClient, EthereumClient {
...
+    const { fast: fastGP, fastest: fastestGP, average: averageGP } = gasPrices
+
+    // gas limits
+    const gasLimits = await this.estimateGasLimits({
```
Comment (veado) [inline packages/xchain-ethereum/src/client.ts:622]: Same here: Handle failing of `estimateGasLimits`

**Review by junkai121 — APPROVED** [review]
LGTM 👍

**Pull request merged**